source/Irrlicht/COpenGLDriver.cpp

```diff
@@ -2236,27 +2236,23 @@ bool COpenGLDriver::SAuxContext::setActiveTexture(uint32_t stage, core::smart_re
 	return true;
 }
 
-void COpenGLDriver::SAuxContext::setActiveBindings(const COpenGLDescriptorSet* descriptorSets[4])
+void COpenGLDriver::SAuxContext::setActiveDescriptorSet(uint32_t index, const COpenGLDescriptorSet* ds)
 {
     //TODO optimize this
-    for (uint32_t i = 0u; i < 4u; ++i)
-    {
-        const COpenGLDescriptorSet* ds = descriptorSets[i];
-        if (!ds)
-            continue;
+    if (!ds)
+        return;
 
-        const auto& multibindParams = ds->getMultibindParams();
-        const auto& multibind_first_count = static_cast<const COpenGLPipelineLayout*>(CurrentRenderpassPipeline->getLayout())->getMultibindParamsForDescSet(i);
+    const auto& multibindParams = ds->getMultibindParams();
```

Comment:
here's a way to do it #25 

Ideally you want to protect every OpenGL call that is likely to happen every frame (so almost all done by COpenGLDriver) with `if (nextValue!=currentValue)` so that you're not pounding OpenGL calls pointlessly.

When you want to change OpenGL state (current pipeline, ROP params, VAO, VAO input buffers, etc. ), dont actually change it straight away, have a deferred change mechanism.

Obviously there are a few exceptions that cannot be done like this, most notably setting framebuffers and push constants (these need to be immediate, not deferred). Sidenote, framebuffer change is a big call with big cost, so users will be careful not to call it recklessly.

Basically you want to avoid a performance penalty when user or implementation sets descriptor bindings to null, and then binds own descriptor sets. Same for VAO, and VAO buffer bindings, etc.
Because what would happen, is first (probably multiple) `extGlBind*(0)` calls before a bind call that actually binds an object.

What you want to do, is to track some of the state (not all OpenGL state) that is relevant to things you're going to be deferring.
So you have a current-state (last flushed) cache, and a next-state cache (values that you will set to if `flushState` is called). Then when your `setActiveDescriptorSet` and similar calls take place, you modify the "next-state" (which holds opengl values you will change to).
This means that while your smart pointers to front irr API objects may change, etc. OpenGL will not be called until it is time. Only when you call `flushState` will you check if next OpenGL values are different from current OpenGL values and call the OpenGL functions if necessary, then update the current state values.

**Don't worry about OpenGL objects being deleted while having their handles on "next-state" thats a user error.**

Implement a `SAuxContext::flushState(SAuxContext::GL_STATE_BITS, COpenGLState* optOutCurrentState)` that will change the current state to next-state and fill out the COpenGLState (already in the engine) class/struct with appropriate data (so it doesn't need to `glGet` query it).

Expose an `IDriver::flushAllState()` that will call `SAuxContext::flushState(SAuxContext::GSB_ALL,nullptr)` under the hood. People won't usually have to call that, as push constants, {multi}dispatch{indirect} and {multi}draw{indirect} will do it under the hood for them. 

My proposal for the flags is (you should also flush/update the OpenGL state in the order of the flags, most costly to least)
```cpp
enum GL_STATE_BITS
{
   // has to be flushed before constants are pushed (before `extGlProgramUniform*`)
   GSB_PIPELINE=0x1u,
   // (all stancil, depth, etc state) change just before draw
   GSB_RASTER_PARAMETERS=0x2u,
   // we want the two to happen together and just before a draw (set VAO first, then binding)
   GSB_VAO_AND_VERTEX_INPUT=0x4u,
   // flush just before (indirect)dispatch or (multi)(indirect)draw, textures and samplers first, then storage image, then SSBO, finally UBO
   GSB_DESCRIPTOR_SETS=0x8u,
   // flush everything
   GSB_ALL=~0x0u
};
```

Also a `SAuxContext::notifyOfState(const COpenGLState& externallyChangedState)` could be useful for interop with middlewares (set the current state cache to that which is reported).

Reply:
`GSB_PIPELINE` and `GSB_RASTER_PARAMETERS` can be replaced by one `GSB_PIPELINE_AND_RASTER_PARAMETERS`

Reply:
@Crisspl unified the enum?

Reply:
yes